keps/sig-network/20191205-kube-proxy-IP-node-binding.md

```diff
@@ -0,0 +1,79 @@
+reviewers:
+  - "@thockin"
+  - "@cheftako"
+  - TBD
+approvers:
+  - "@thockin"
+  - TBD
+editor: TBD
+creation-date: 2019-12-05
+last-updated: 2019-12-21
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Give the ability to the LoadBalancer provisioner or a user to disable the LB IP binding
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+<!-- /toc -->
+
+## Summary
+
+The different kube-proxy implementations (at least ipvs and iptables), as of today, are binding the External IPs of the LoadBalancer Service to each node. (iptables are creating some rules to redirect packets directly to the service and ipvs is binding the IP to one interface on the node). This feature exists because:
+
+1. To redirect pod traffic that is intended for LoadBalancer IP directly to the service, bypassing the loadbalancer for path optimization.
+2. Some LBs are sending packet with the destination IP equal to the LB IP, so the IP needs to be routed directly to the right service (otherwise looping back and forth)
+3. (TBD)
+
+This enhancement proposes a way to make these features configurable, giving the cloud provider, or the user, a way to disable the behaviour of kube-proxy.
+The best way would be to set a default behaviour at the cloud controller level, that can be overwritten by the user at a service level.
+
+## Motivation
+
+There are numerous problems with the current behaviour:
+
+1. Some cloud providers (Scaleway, Tencent Cloud, ...) are sometimes using the LB's external IP as source IP when sending packets to the cluster. This is a problem in the ipvs mode of kube-proxy since the IP is bounded to an interface and healthchecks from the LB are never coming back.
+2. Some cloud providers (DigitalOcean, Scaleway, ...) have features at the LB level (TLS termination, PROXY protocol, ...). Bypassing the LB means missing these features when the packet arrives at the service (leading to protocols errors).
+
+So, giving options to these cloud providers to disable the actual behaviour would be very valuable.
+
+Currently there is some hacky workaround that set the `Hostname` on the `Service` in order to bypass kube-proxy binding (AWS and DigitalOcean for instance), but as stated, it's a bit hacky. This KEP should also give a solution to this hack.
+
+### Goals
+
+* Expose an option for the different controllers managing Load Balancer (e.g. the Cloud Controller Manager), to disable the actual behaviour of kube-proxy by default for all services.
+
+### Non-Goals
+
+* Deprecate the default behaviour of kube-proxy
+* Change the default behaviour of kube-proxy
+
+## Proposal
+
+The solution would be to add a new field in the `LoadBalancerStatus` field of a `Service`, like `BindLocal`. This new field will be used by kube-proxy in order to not bind the Load Balancer's External IP to the node (in both IPVS and iptables mode).
```

Comment:
Just to be clear - when you say "bind", you don't mean `bind()` or `ip addr add`, but the specific iptables rules used to capture local traffic, right?

Reply:
Yep I meant the iptables rules that catches the LB IP. (Though with ipvs it does an `ip addr add`, so this one will disappear too)